crates/ra_syntax/src/validation.rs

```diff
@@ -1,40 +1,89 @@
 use crate::{
+    algo::visit::{visitor_ctx, VisitorCtx},
     ast::{self, AstNode},
     File,
-    string_lexing,
+    string_lexing::{self, CharComponentKind},
     yellow::{
         SyntaxError,
     },
 };
 
 pub(crate) fn validate(file: &File) -> Vec<SyntaxError> {
     let mut errors = Vec::new();
-    for d in file.root.borrowed().descendants() {
-        if let Some(c) = ast::Char::cast(d) {
-            let components = &mut string_lexing::parse_char_literal(c.text());
-            let len = components.count();
+    let _ = visitor_ctx(&mut errors)
+        .visit::<ast::Char, _>(validate_char)
+        .accept(file.root.borrowed());
```

Comment:
Visitor is not recursive, you still have to drive traversal yourself, via iterating descendants.